llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h

```diff
@@ -111,21 +148,43 @@ class TransformInterpreterPassBase : public GeneratedBase<Concrete> {
   /// fails.
   LogicalResult runAfterInterpreter(Operation *) { return success(); }
 
+  /// Hook for passes to indicate payload roots for successive invocations of
+  /// the transform dialect interpreter.
+  void getPayloadRoots(SmallVectorImpl<Operation *> &targets) {
+    auto *pass = static_cast<Concrete *>(this);
+    targets.push_back(pass->getOperation());
+  }
+
+  /// Hook for passes to construct transform IR separately from payload IR.
+  OwningOpRef<ModuleOp> constructTransformModule(MLIRContext *) {
+    return nullptr;
+  }
+
   void runOnOperation() final {
     auto *pass = static_cast<Concrete *>(this);
-    Operation *op = pass->getOperation();
-    if (failed(pass->runBeforeInterpreter(op)) ||
-        failed(detail::interpreterBaseRunOnOperationImpl(
-            op, pass->getArgument(), sharedTransformModule,
-            pass->transformFileName, pass->debugPayloadRootTag,
-            pass->debugTransformRootTag)) ||
-        failed(pass->runAfterInterpreter(op))) {
-      return pass->signalPassFailure();
+    SmallVector<Operation *> payloadRoots;
+    pass->getPayloadRoots(payloadRoots);
+    for (Operation *root : payloadRoots) {
+      if (failed(pass->runBeforeInterpreter(root)) ||
+          failed(detail::interpreterBaseRunOnOperationImpl(
+              root, pass->getArgument(), sharedTransformModule,
+              pass->transformFileName, pass->debugPayloadRootTag,
+              pass->debugTransformRootTag)) ||
+          failed(pass->runAfterInterpreter(root))) {
+        return pass->signalPassFailure();
+      }
     }
   }
 
+protected:
+  /// Returns `true` if the separate transformation module is available as a
+  /// result of parsing or construction.
```

Comment:
This comment prob needs some more explanation, I am not able to link parsing or construction to the rest of the documentation in this file at first glance.

Reply:
Added more information above, here is the wrong place.